taar_lite/recommenders/guid_based_recommender.py

```diff
@@ -55,13 +114,72 @@ def recommend(self, client_data, limit=4):
         TAAR lite will yield 4 recommendations for the AMO page
         """
         addon_guid = client_data.get('guid')
-        result_dict = self.addons_coinstallations.get(addon_guid, {})
-        result_list = list(result_dict.items())
+        normalize = client_data.get('normalize', None)
+        norm_dict = {'uniq_coinstall': self.norm_uniq_coinstall,
+                     'rowsum': self.norm_rowsum,
+                     'rownorm_sum': self.norm_rownorm_sum}
 
-        result_list = sorted(result_list, key=lambda x: x[1], reverse=True)
+        if normalize is not None and normalize not in norm_dict.keys():
+            # Yield no results if the normalization method is not
+            # specified
+            logger.warn("Invalid normalization parameter detected: [%s]" % normalize)
+            return []
 
-        # TODO: normalize confidence output based on frequncy divided by total
-        # frequency sum of all addon installations observed.
-        # TODO: replace '1.0' with the normalized relative frequency from above
+        result_dict = self._addons_coinstallations.get(addon_guid, {})
+
+        # Default the normalization method to no normalization
+        norm_method = norm_dict.get(normalize, lambda guid, x: x)
+        result_dict = norm_method(addon_guid, result_dict)
+
+        result_list = sorted(result_dict.items(), key=lambda x: x[1], reverse=True)
 
         return result_list[:limit]
+
+    def norm_uniq_coinstall(self, key_guid, input_coinstall_dict):
+        """This normalization method counts the unique times that a
+        GUID is coinstalled with any other GUID.
+
+        This dampens weight of any suggested GUID inversely
+        proportional to it's overall popularity.
+        """
+        uniq_guid_map = self._guid_maps['unique_coinstall']
+
+        output_result_dict = {}
+        for result_guid, result_count in input_coinstall_dict.items():
+            output_result_dict[result_guid] = 1.0 * result_count / uniq_guid_map[result_guid]
+        return output_result_dict
+
+    def norm_rowsum(self, key_guid, input_coinstall_dict):
+        """This normalization normalizes the weights for the suggested
+        coinstallation GUIDs based on the sum of the weights for the
+        coinstallation GUIDs given a key GUID.
+        """
+        guid_count_map = self._guid_maps['count_map']
+        return dict([(k, v*1.0/guid_count_map[k])
+                    for k, v
+                    in input_coinstall_dict.items()])
+
+    def norm_rownorm_sum(self, key_guid, input_coinstall_dict):
+        """This normalization is the same as norm_rowsum, but we also
+        divide the dampen by the total number of coinstallations.
+        """
+
+        # Compute an intermediary dictionary that is the result dict
+        # normalizing each coinstalled GUID by it's occurance in the
+        # entire space
```

Comment:
Clarification: does this mean its total co-installation count with all other guids?

Reply:
Yes - updated this comment in the doc.